### PR TITLE
feat: handle moderate price impact at input/confirm step

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Does what it says on the box - if the price impact is greater than 5%, adds a warning both at input and confirm steps, and changes the confirm button to an errored "Trade Anyway" one.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/4856

## Risk

We may show this when we shouldn't, test accordingly

## Testing

- Refer to the testing steps of https://github.com/shapeshift/web/pull/5032, except instead of a window confirm, this will actually be visible at input and confirm steps

### Engineering

- ☝🏽 

### Operations

- ☝🏽 

## Screenshots (if applicable)

image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5032** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5032" target="_blank">Graphite</a> 
    * **PR #5033** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5033" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5033?utm_source=stack-comment).

File: src/components/MultiHopTrade/components/PriceImpact.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        <FaBolt color='red.500' />
+        <Text translation='trade.priceImpact' />
+      </Row.Label>
+      <Row.Value>
+        <HelperTooltip
+          label={translate('trade.tooltip.priceImpact')}
+          flexProps={{ flexDirection: 'row' }}
+          iconProps={{ color: redText }}
+        >
+          <Text translation={`${impactPercentage} %`} color={redText} />
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/PriceImpact.tsx:29]: CText

File: src/components/MultiHopTrade/components/TradeConfirm/TradeConfirm.tsx [hunk 2]
```diff
@@ -402,20 +416,22 @@ export const TradeConfirm = () => {
...
+              colorScheme={isModeratePriceImpact ? 'red' : 'blue'}
               size='lg'
               width='full'
               mt={6}
               data-test='trade-form-confirm-and-trade-button'
               type='submit'
             >
-              <Text translation='trade.confirmAndTrade' />
+              <Text
+                translation={isModeratePriceImpact ? 'Trade anyway' : 'trade.confirmAndTrade'}
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeConfirm/TradeConfirm.tsx:427]: translation

Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeConfirm/TradeConfirm.tsx]: e722ecd4214865eef34c06e7152e72783cda9ab1

File: src/components/MultiHopTrade/components/PriceImpact.tsx (hunk 1 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/PriceImpact.tsx]: e722ecd4214865eef34c06e7152e72783cda9ab1

File: src/components/MultiHopTrade/components/PriceImpact.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        <Text translation='trade.priceImpact' />
+        <RawText as='span' fontSize='lg' color='gray.500'>
+          <MdOfflineBolt />
+        </RawText>
+      </Row.Label>
+      <Row.Value>
+        <HelperTooltip
+          label={translate('trade.tooltip.priceImpact')}
+          flexProps={{ flexDirection: 'row' }}
+          iconProps={{ color: redText }}
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/PriceImpact.tsx:29]: Extract me (props object) for render performance improvements.

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx [hunk 4]
```diff
@@ -104,7 +106,24 @@ export const TradeInput = memo(() => {
...
   const sellAmountCryptoPrecision = useAppSelector(selectSellAmountCryptoPrecision)
-  const slippagePercentage = useAppSelector(selectTradeSlippagePercentageDecimal)
+  const sellAmountBeforeFeesUserCurrency = useAppSelector(selectSellAmountUserCurrency)
+  const slippageDecimal = useAppSelector(selectTradeSlippagePercentageDecimal)
+
+  const priceImpactPercentage = useMemo(() => {
+    if (!sellAmountBeforeFeesUserCurrency || !buyAmountAfterFeesUserCurrency) return bn('0')
+
+    const tradeDifference = bn(sellAmountBeforeFeesUserCurrency)
+      .minus(buyAmountAfterFeesUserCurrency)
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:116]: As with the previous PR, it feels kind of strange to be using the after fees amount here, as they shouldn't impact the messaging we give users related to pool liquidity.

Comment (0xApotheosis) [reply src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx:117]: https://github.com/shapeshift/web/pull/5033/commits/40e7adaa2388993eff19afbb4887fbc26644e98c

File: src/components/MultiHopTrade/components/PriceImpact.tsx (hunk 3 above)
Comment (0xApotheosis) [reply src/components/MultiHopTrade/components/PriceImpact.tsx]: https://github.com/shapeshift/web/pull/5033/commits/40e7adaa2388993eff19afbb4887fbc26644e98c

**Review by 0xApotheosis — APPROVED** [review]
Indeed does what it says on the box. Tested the entire stack:

Screenshot 2023-08-09 at 2 36 03 pm
Screenshot 2023-08-09 at 2 36 14 pm
Screenshot 2023-08-09 at 2 36 20 pm
Screenshot 2023-08-09 at 2 37 36 pm

File: src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx (hunk 4 above)
Comment (0xApotheosis) [reply src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx]: https://github.com/shapeshift/web/pull/5033/commits/ae775300ec56af378f0225a167859f62ea050e32

**Pull request merged**